File: python/dominoes/dominoes.py
from itertools import permutations


def can_chain(dominoes):
    if dominoes == []:
        return []

    for permutation in permutations(dominoes):
        chain = [permutation[0]]

        for x, y in permutation[1:]:
            z = chain[-1][1]

            if x == z:
                chain.append((x, y))
            elif y == z:
                chain.append((y, x))
            else:
                break

        if len(chain) == len(permutation) and chain[0][0] == chain[-1][1]:
            return chain

    return None

File: haskell/dominoes/src/Dominoes.hs
module Dominoes
  ( chain
  ) where

import           Data.List                      ( find
                                                , foldl'
                                                , permutations
                                                )

chain :: [(Int, Int)] -> Maybe [(Int, Int)]
chain xs = find isValidChain (map chain' (permutations xs))
 where
  isValidChain [] = True
  isValidChain ys = length xs == length ys && fst (head ys) == snd (last ys)

chain' :: [(Int, Int)] -> [(Int, Int)]
chain' []            = []
chain' ((a, b) : xs) = foldl' f [(a, b)] xs
 where
  f [] _ = []
  f xs'@((a', _) : _) (c, d) | a' == c   = (d, c) : xs'
                             | a' == d   = (c, d) : xs'
                             | otherwise = xs'
